File: src/test/java/ru/xpendence/topjavagraduation/DataBuilder.java
package ru.xpendence.topjavagraduation;

import org.apache.commons.lang3.RandomStringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import ru.xpendence.topjavagraduation.entity.Dish;
import ru.xpendence.topjavagraduation.entity.Restaurant;
import ru.xpendence.topjavagraduation.entity.Review;
import ru.xpendence.topjavagraduation.entity.Tag;
import ru.xpendence.topjavagraduation.entity.User;
import ru.xpendence.topjavagraduation.entity.Vote;
import ru.xpendence.topjavagraduation.repository.DishRepository;
import ru.xpendence.topjavagraduation.repository.RestaurantRepository;
import ru.xpendence.topjavagraduation.repository.ReviewRepository;
import ru.xpendence.topjavagraduation.repository.TagRepository;
import ru.xpendence.topjavagraduation.repository.UserRepository;
import ru.xpendence.topjavagraduation.repository.VoteRepository;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.Random;

@Component
public class DataBuilder {

    @Autowired
    private RestaurantRepository restaurantRepository;

    @Autowired
    private DishRepository dishRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private VoteRepository voteRepository;

    @Autowired
    private ReviewRepository reviewRepository;

    @Autowired
    private TagRepository tagRepository;

    private final Random RANDOM = new Random();

    public Restaurant buildRestaurant() {
        var restaurant = new Restaurant();
        restaurant.setName(RandomStringUtils.secure().nextAlphanumeric(16));
        return restaurant;
    }

    public Restaurant saveRestaurant() {
        return restaurantRepository.save(buildRestaurant());
    }

    public Dish buildDish(Restaurant restaurant) {
        var dish = new Dish();
        dish.setName(RandomStringUtils.secure().nextAlphanumeric(16));
        dish.setPrice(BigDecimal.valueOf(RANDOM.nextDouble()));
        dish.setActive(RANDOM.nextBoolean());
        dish.setRestaurant(restaurant);
        return dish;
    }

    public Dish saveDish(Restaurant restaurant) {
        return dishRepository.save(buildDish(restaurant));
    }

    public User buildUser() {
        var user = new User();
        user.setUsername(RandomStringUtils.secure().nextAlphanumeric(16));
        user.setPassword(RandomStringUtils.secure().nextAlphanumeric(16));
        return user;
    }

    public User saveUser() {
        return userRepository.save(buildUser());
    }

    public Vote buildVote(User user, Restaurant restaurant) {
        return buildVote(user, restaurant, LocalDate.now());
    }

    public Vote buildVote(User user, Restaurant restaurant, LocalDate date) {
        var vote = new Vote();
        vote.setDate(date);
        vote.setUser(user);
        vote.setRestaurant(restaurant);
        return vote;
    }

    public Vote saveVote(User user, Restaurant restaurant) {
        return voteRepository.save(buildVote(user, restaurant));
    }

    public Vote saveVote(User user, Restaurant restaurant, LocalDate date) {
        return voteRepository.save(buildVote(user, restaurant, date));
    }

    public Review buildReview(User user, Restaurant restaurant) {
        return buildReview(user, restaurant, 5, RandomStringUtils.secure().nextAlphanumeric(32));
    }

    public Review buildReview(User user, Restaurant restaurant, Integer rating, String text) {
        var review = new Review();
        review.setRating(rating);
        review.setText(text);
        review.setDate(LocalDate.now());
        review.setUser(user);
        review.setRestaurant(restaurant);
        return review;
    }

    public Review saveReview(User user, Restaurant restaurant) {
        return reviewRepository.save(buildReview(user, restaurant));
    }

    public Review saveReview(User user, Restaurant restaurant, Integer rating, String text) {
        return reviewRepository.save(buildReview(user, restaurant, rating, text));
    }

    public Tag buildTag() {
        var tag = new Tag();
        tag.setName(RandomStringUtils.secure().nextAlphanumeric(16));
        return tag;
    }

    public Tag saveTag() {
        return tagRepository.save(buildTag());
    }
}

File: src/test/kotlin/ru/xpendence/topjavagraduation/DataBuilder.kt
package ru.xpendence.topjavagraduation

import org.apache.commons.lang3.RandomStringUtils
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.stereotype.Component
import ru.xpendence.topjavagraduation.entity.*
import ru.xpendence.topjavagraduation.repository.*
import java.math.BigDecimal
import java.time.LocalDate
import java.util.*

@Component
class DataBuilder(
    private val restaurantRepository: RestaurantRepository,
    private val dishRepository: DishRepository,
    private val userRepository: UserRepository,
    private val voteRepository: VoteRepository,
    private val reviewRepository: ReviewRepository,
    private val tagRepository: TagRepository,
    private val jdbcTemplate: JdbcTemplate,
) {
    private val random = Random()

    fun buildRestaurant(): Restaurant =
        Restaurant(name = RandomStringUtils.secure().nextAlphanumeric(16))

    fun saveRestaurant(): Restaurant =
        restaurantRepository.save(buildRestaurant())

    fun buildDish(restaurant: Restaurant): Dish =
        Dish(
            name = RandomStringUtils.secure().nextAlphanumeric(16),
            price = BigDecimal.valueOf(random.nextDouble()),
            active = random.nextBoolean(),
            restaurant = restaurant,
        )

    fun saveDish(restaurant: Restaurant): Dish =
        dishRepository.save(buildDish(restaurant))

    fun saveDish(restaurant: Restaurant, active: Boolean): Dish =
        dishRepository.save(buildDish(restaurant).copy(active = active))

    fun buildUser(): User =
        User(
            username = RandomStringUtils.secure().nextAlphanumeric(16),
            password = RandomStringUtils.secure().nextAlphanumeric(16),
        )

    fun saveUser(): User =
        userRepository.save(buildUser())

    fun buildVote(user: User, restaurant: Restaurant): Vote =
        buildVote(user, restaurant, LocalDate.now())

    fun buildVote(user: User, restaurant: Restaurant, date: LocalDate): Vote =
        Vote(date = date, user = user, restaurant = restaurant)

    fun saveVote(user: User, restaurant: Restaurant): Vote =
        voteRepository.save(buildVote(user, restaurant))

    fun saveVote(user: User, restaurant: Restaurant, date: LocalDate): Vote =
        voteRepository.save(buildVote(user, restaurant, date))

    fun clearVotes() {
        jdbcTemplate.update("DELETE FROM votes")
    }

    fun buildReview(user: User, restaurant: Restaurant): Review =
        buildReview(user, restaurant, 5, RandomStringUtils.secure().nextAlphanumeric(32))

    fun buildReview(user: User, restaurant: Restaurant, rating: Int, text: String): Review =
        Review(
            rating = rating,
            text = text,
            date = LocalDate.now(),
            user = user,
            restaurant = restaurant,
        )

    fun saveReview(user: User, restaurant: Restaurant): Review =
        reviewRepository.save(buildReview(user, restaurant))

    fun saveReview(user: User, restaurant: Restaurant, rating: Int, text: String): Review =
        reviewRepository.save(buildReview(user, restaurant, rating, text))

    fun buildTag(): Tag =
        Tag(name = RandomStringUtils.secure().nextAlphanumeric(16))

    fun saveTag(): Tag =
        tagRepository.save(buildTag())
}
